File: 2022/AoC_4/day_4.py
with open("input_ex.txt") as file:
    data = file.read().split("\n")


lines = [lines.split(',') for lines in data]
# print(lines)
data_with_ranges = []

# Converting string pairs into range of numbers
for pairs in lines:
    pairs_ranges = []
    for pair in pairs:
        start, end = pair.split("-")
        pairs_ranges.append(range(int(start),int(end) +1))

    # print(pairs_ranges)
    data_with_ranges.append(pairs_ranges)


print(data_with_ranges)

sum1 = 0
sum2 = 0
for pairs in data_with_ranges:
    print(pairs)
    
    section1 = set(pairs[0])
    section2 = set(pairs[1])
    if section1.issubset(section2) or section2.issubset(section1):
            sum1+=1
    if section1.intersection(section2):
         sum2 +=1
         

print(sum1)
print(sum2)

File: 2022/AoC_4/day_4.js
const fs = require('fs');


const fileContent = fs.readFileSync('input_ex.txt', 'utf8');
const data = fileContent.split('\n'); 

console.log(data)

let dataRanges = [];

data.forEach(line => {
    let pairs = line.split(",");
    let pairsRange = []
    pairs.forEach(pair => {
        let [start, end] = pair.split("-").map(Number)
        let range = [];
            for(i = start; i <=end ; i++) {
                range.push(i)
            }
        pairsRange.push(range)
    })
    // console.log(pair)
    dataRanges.push(pairsRange)
})

console.log(dataRanges)


// Function to chech two section and take all elements from arr to check if they are in second arr.
const isSectioninSection = (section1,section2) => {
    return section1.every(element => section2.includes(element));
}

const isSomeSectioninSection = (section1,section2) => {
    return section1.some(element => section2.includes(element));
}

// filter?
// set?



let ans1 = 0;
let ans2 = 0;
dataRanges.forEach(pair => {
    section1 = pair[0];
    section2 = pair[1];
    // wrong because i have to use single element - not whole array to check
    // if (section1.inludes(section2) || section2.includes(section1)){
    //     ans1 += 1
    // }
    if (isSectioninSection(section1,section2) || isSectioninSection(section2,section1)){
        ans1 +=1
    }
    if (isSomeSectioninSection(section1,section2) || isSomeSectioninSection(section2,section1)){
        ans2 +=1
    }
})

console.log(ans1)
console.log(ans2)